### ISSUE TITLE
Images with inline source are not working data:image/png;base64 

### ISSUE THREAD

**Issue by Rusu-Marian-Dan** [richard-scryber/scryber.core#45]
Tested both inside header and inside body

**Comment by Rusu-Marian-Dan** [comment]
Looking into code, \Scryber.Components\Components\Image.cs
I think it is expected that the image src is a full path and also is adding the resulting xobj into SharedResources (obviously, base64 will not be able to be added to SharedResources as it has no way to identify it later)

**Comment by richard-scryber** [comment]
Thanks for the input, and no - inline images are not currently supported. Would not be a massive effort though, and I can put it in the backlog. 

There are image factories that could be added to the to the configuration options, for the data: path
I'll see if I can come up with an example.

**Comment by Rusu-Marian-Dan** [comment]
If there are some tooling that must be used, please point me to it, I can help (eg: a class that you think must be extended to support the inline source)
I can just do the change in place, but I do not understand enough about your architecture to say if it is the correct way
Thanks

**Comment by richard-scryber** [comment]
I created a basic image factory implementing the IPDFImageDataFactory

``` csharp

        public class DataBase64ImageFactory : IPDFImageDataFactory
        {

            public bool ShouldCache { get { return false; } }

            /// <summary>
            /// Loads an image from a base 64 data image
            /// </summary>
            /// <param name="document"></param>
            /// <param name="owner"></param>
            /// <param name="path"></param>
            /// <returns></returns>
            public Scryber.Drawing.PDFImageData LoadImageData(IPDFDocument document, IPDFComponent owner, string path)
            {
                path = GetBase64FromPath(path);

                var binary = Convert.FromBase64String(path);

                using (var ms = new System.IO.MemoryStream(binary))
                {
                    return PDFImageData.LoadImageFromStream(document.GetIncrementID(owner.Type) + "data_png", ms, owner);
                }
            }

            /// <summary>
            /// Strips the front from the image path and returns the base64 string
            /// </summary>
            /// <param name="path"></param>
            /// <returns></returns>
            private static string GetBase64FromPath(string path)
            {
                path = path.Trim();
                var index = path.IndexOf("data:image/png;");

                if (index <0)
                    throw new NotSupportedException("This image factory only supports data png images");

                //could handle different types of image not just png
                path = path.Substring(index + "data:image/png;".Length);

                if (!path.StartsWith("base64,"))
                    throw new NotSupportedException("This image factory only supports base64 encoded data images");

                path = path.Substring("base64,".Length);
                path = path.TrimStart();

                //TODO: remove any carriage returns etc.

                return path;
            }
        }
```

This can then be added to the document image factories based on path matching, or in the settings.json file

``` csharp
        [TestMethod]
        public void DataImageFactory_Test()
        {
             var html = @"<html xmlns='http://www.w3.org/1999/xhtml' ><body style='padding:20pt;' >
<img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==' alt='Red dot' />
</body></html>";

            using (var sr = new System.IO.StringReader(html))
            {
                using (var doc = Document.ParseDocument(sr, ParseSourceType.DynamicContent))
                {
                    // Can add in the settings.json, or manually here.
                    EnsureDataImageFactory(doc);

                    using (var stream = DocStreams.GetOutputStream("DataImage.pdf"))
                    {
                        doc.SaveAsPDF(stream);
                    }

                }
            }

        }

        public void EnsureDataImageFactory(Document doc)
        {
            var factories = doc.ImageFactories;

            if (factories.Count == 0 || factories.FirstOrDefault(f => { return f.Name == "DataImages"; }) == null)
            {
                var dataImg = new Options.PDFImageFactory("DataImages", new Regex("data:"), new DataBase64ImageFactory());
                factories.Add(dataImg);
            }
        }
```

The factory needs a clean up to support other types and returns in the src attribute (https://en.wikipedia.org/wiki/Data_URI_scheme), but might work for you in the mean time.

**Comment by richard-scryber** [comment]
If you do manage to get a good implementation would happily pull it back into the source! :-)

**Comment by Rusu-Marian-Dan** [comment]
This is really helpful, It will unblock me immediately.

I have forked it and added a permanent fix (since the URI is verified to be valid in LoadImageData method, we can also check if it is a "data" scheme, then use PDFImageData.LoadImageFromUriData (new method) to load it's content)

I think the change is not too intrusive.

Implementation of PDFImageData.LoadImageFromUriData verifies mime and encoding and loads content into PDFImageData

Let me know if you think it wort pulling it to your repo and the branch where

Thanks
Dan

**Comment by richard-scryber** [comment]
I have added a new dataImages branch, so we can check the changes. Really appreciate the input Dan

**Comment by richard-scryber** [comment]
Added to the main source Dan, thank you. It will come out in the package in the next week or so, and I will close then. :-)

**Comment by richard-scryber** [comment]
The multiple images with the same path had got messed up anyway, so I have put a fix and a test in for that too.

**Comment by Rusu-Marian-Dan** [comment]
Happy to help

**Comment by richard-scryber** [comment]
@Rusu-Marian-Dan - released latest version 5.0.6 that supports inline images. Closing this, and thanks for the input

**Comment by Naliwe** [comment]
Hello :)

Sorry to re-open this, but I found myself in the same hussle and it seems like what you've merged only takes `png` into account; which is a bit sad :/

Do you confirm that's the case ? My tests seem to indicate I can only generate images from `data:image/png;base64, ...`.

I tried using your same class that's been posted here, removed the check on `/png` and moved the index to include everyting before the `,`, as this will always be the case no matter the image actual format. The renderer should handle any errors that should arise if the format happens to be incorrect.

I'm using a modified version of the code provided here now, but I'd like to be able to use the framework itself and not some custom piece of code if possible :p What do you think about extending the supported image formats in your code?
